### PR TITLE
Add StructureModifier for extracting the signature data in chat and login packets

### PR DESCRIPTION
**Pull request by games647**
Related #1661

Here is the changes for the linked ticket. I'm not that familiar with the internals of ProtocolLib, so I would appreciate any feedback. 

## Summary of changes

I implemented the changes by wrapping the Either class of Mojang with a simple one of our own. The converting step however requires the dependency on the datafixers class, but this is included in any Spigot server implementation.

## Checklist

- [X] Either converter
- [X] Wrapper
  - [X] LoginSignature
  - [X] Either
- [X] StructureModifer for chat messages
- [X] Tests
- [x] Documentation

## Questions

As you can see there are huge file changes. This happens, because the previous files had windows line endings, which were automatically converted by my Git installation to Linux ones. I could keep the existing ones, but most of the remaining project already uses the Linux line endings too.

The documentation is currently missing, because I would like to get feedback for the suggested changes first.

### CODE REVIEW

**Review by derklaro — COMMENTED** [review]
I have to say that it's pretty hard to review this if you reformat all classes on the fly...

**Comment by games647** [comment]
@derklaro Could you check the newest version if you have some time? I updated the code to remove all formatting changes.

File: src/main/java/com/comphenix/protocol/utility/MinecraftReflection.java [hunk 1]
```diff
@@ -2274,6 +2273,10 @@ public static Class<?> getProfilePublicKeyClass() {
 		return getMinecraftClass("world.entity.player.ProfilePublicKey");
 	}
 
+    public static Class<?> getLoginSignatureClass() {
+        return getMinecraftClass("util.MinecraftEncryption").getClasses()[0];
```
Comment (dmulloy2) [inline src/main/java/com/comphenix/protocol/utility/MinecraftReflection.java:2277]: I don't love relying on the order here; does this class have any other identifying characteristics we could use?

File: src/main/java/com/comphenix/protocol/wrappers/BukkitConverters.java [hunk 2]
```diff
@@ -564,6 +596,10 @@ public static EquivalentConverter<WrappedProfileKeyData> getWrappedPublicKeyData
 		return ignoreNull(handle(WrappedProfileKeyData::getHandle, WrappedProfileKeyData::new, WrappedProfileKeyData.class));
 	}
 
+    public static EquivalentConverter<WrappedLoginSignature> getWrappedSignatureConverter() {
```
Comment (dmulloy2) [inline src/main/java/com/comphenix/protocol/wrappers/BukkitConverters.java:599]: indentation here and a few other places looks a little off. could just be the web editor being weird though

File: src/main/java/com/comphenix/protocol/wrappers/Either.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
+package com.comphenix.protocol.wrappers;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+public abstract class Either<L, R> {
+
+    public static class Left<L, R> extends Either<L, R> {
```
Comment (dmulloy2) [inline src/main/java/com/comphenix/protocol/wrappers/Either.java:8]: is there a reason for having them be separate classes vs a single class? i'd imagine you could get the same functionality in one class with an exclusive or

Comment (dmulloy2) [reply src/main/java/com/comphenix/protocol/wrappers/Either.java:8]: could have static factory methods `left` and `right`

**Review by dmulloy2 — CHANGES_REQUESTED** [review]

File: src/main/java/com/comphenix/protocol/wrappers/Either.java (hunk 3 above)
Comment (games647) [reply src/main/java/com/comphenix/protocol/wrappers/Either.java:8]: >could have static factory methods left and right

I'll change that.

>is there a reason for having them be separate classes vs a single class? i'd imagine you could get the same functionality in one class with an exclusive or

The implementation is similar to Mojang's implementation. Having separate classes allows us to have a single field while also being type-safe. Left and Right could be/are different types. Therefore we either use generic Object or two fields, where one them is null.

Pseudo:
```java
class Either {

private final Object value;
private boolean left;

public Optional<L> left() {
    return left ? Optional.empty() : Optional.of((L) value);
}

public Optional<R> right() {
    return left ? Optional.of((R) value) : Optional.empty();
}
```

or

```java
class Either {

private final L left;
private final R right;

public Optional<L> left() {
    return Optional.ofNullable(left);
}

public Optional<R> right() {
    return Optional.ofNullable(right);
}
```
Note: The constructor could then provide the exclusivity.

File: src/main/java/com/comphenix/protocol/wrappers/BukkitConverters.java (hunk 2 above)
Comment (games647) [reply src/main/java/com/comphenix/protocol/wrappers/BukkitConverters.java:599]: Well in this case it's same as the other methods. However, I think I know what you mean. I adopted the checkstyle format, so you might noticed the following breaks if the line is too long.

```java
public void tooLong(String param1,
                    String param2) {
```

or 

```java
doSomething(
    param1,
    param2
);
```

Should I remove that?

File: src/main/java/com/comphenix/protocol/utility/MinecraftReflection.java (hunk 1 above)
Comment (games647) [reply src/main/java/com/comphenix/protocol/utility/MinecraftReflection.java:2277]: We could extract the type from the packet classes, but this seems unreliable in the future. There also the idea to scan all subclasses of Minecraft Reflection to get the correct constructor. Furthermore there is a class for signing chat messages. There we could use fuzzy reflection contract to extract the correct field with the expected constructor:

```suggestion
		Class<?> messageSigClass = getMinecraftClass("network.chat.MessageSignature", "MessageSignature");

		FuzzyClassContract signatureContract = FuzzyClassContract.newBuilder().
				constructor(FuzzyMethodContract.newBuilder().
						parameterCount(2).
						parameterSuperOf(Long.TYPE, 0).
						parameterSuperOf(byte[].class, 1).
						build()
				).build();

		FuzzyFieldContract fuzzyFieldContract = FuzzyFieldContract.newBuilder().
				typeMatches(signatureContract).
				build();

		return FuzzyReflection.fromClass(messageSigClass, true).getField(fuzzyFieldContract).getType();
```

**Comment by games647** [comment]
I noticed that the signature struct is also used for chat messages. Therefore I added two addition StructureModifier methods (One generic `getEithers` and one specific for chat messages `getSignatures`). Furthermore I renamed the Wrapper to reflect that it's not only used for login verification.

File: src/main/java/com/comphenix/protocol/wrappers/Either.java (hunk 3 above)
Comment (dmulloy2) [reply src/main/java/com/comphenix/protocol/wrappers/Either.java:8]: I'm good with leaving them as separate classes if that's what Mojang does. Might still be worth having those factory methods in Either

File: src/main/java/com/comphenix/protocol/utility/MinecraftReflection.java (hunk 1 above)
Comment (dmulloy2) [reply src/main/java/com/comphenix/protocol/utility/MinecraftReflection.java:2277]: I would prefer using fuzzy reflection. It only needs to be run once, then the result gets cached

**Review by dmulloy2 — DISMISSED** [review]

**Comment by dmulloy2** [comment]
Looks good to me. There are some failing tests that seem to be a result of some other recently merged PRs

**Comment by games647** [comment]
I could bisect the bug back to the MethodHandles pull request. The MinecraftObjectMatcher doesn't work anymore and it's related to the change in the [Java regex](https://github.com/dmulloy2/ProtocolLib/blame/master/src/main/java/com/comphenix/protocol/utility/MinecraftReflection.java#L61). Changing it back fixes the test. 

Since that change is part of big commit (maybe squashed?), I cannot extract further details that motivated this change. @derklaro Could you maybe take a look, because you contributed this change.

**Review by dmulloy2 — APPROVED** [review]

**Pull request merged**